# Capstone Project - Final
### Weeks 4 & 5

## Table of contents
* [Introduction: Business Problem](#business_problem)
* [Data](#data)
* [Methodology](#methodology)
* [Analysis](#analysis)
* [Results and Discussion](#results)
* [Conclusion](#conclusion)

## Business Problem <a name="business_problem"></a>

The city of Ljubljana, capital of Slovenia, has been a very popular tourist destination in the recent  years and it hosts more tourists each year. Also more and more tourists decide to stay in Ljubljana for more than 1 day.  Thus the accommodation offering (hotels, motels, inns, bed & breakfasts etc. ) needs to grow with the demand.  

One of the most important factors when looking for a location for tourist accommodation is location. The goal of this project is to identify parts of the city of Ljubljana  that could be candidates for location for tourist acommodation. 

The report is targeting investors/stakeholders/contractors that are interested in creating new accommodation offerings in Ljubljana. The results should be interesting for either small sized (butique) acommodations as well as higher capacity accommodation.  

There are other factors that impact the decision which are not considered in this project, like realestate prices. 

## Data <a name="data"></a>

The following assumptions will be considered when trying to find a potential location of tourist accommodation:
* No other tourist accommodation venue in vicinity. 
* Gravitating towards the city center. Most of Ljubljana's attractions are in the city center. 
* Walking distance away to existing food & drink venues. We will not distinct specifically on the type of food venue. 
* Walking distance to the public transportation grid (bus stations/stops).

The main  datasource will be Foursquare database for the following information:
* Location of existing accommodation venues (any type). 
* Location of existing food venues (any type).
* Bus station/stop locations. 

Transformations between addresses and geographic locations will be done using the Nomatim API. 

Locations of administrative regions (neighbourhoods/boroughs/districts) could be used as source points for Foursquare requests. However those regions have incorrect shapes and very different sizes which can skew the data. To avoid search gaps it was decided to place a square grid over the city of Ljubljana, spanning several kilometers and centered at the most popular tourist spot, "Ljubljansko tromostovje".
Each grid unit/area will be used as a bounding box for retrieving venues from Foursquare using the search API.


## Methodology

The goal is to finds part of the city of Ljubljana that have low accomodation density. At the same time those areas need to have close (walking distance) access to the public transportation grid and also several food venues. 

In the first steps we collect the venue information from Foursquare from an area of several kilometers. To ease the data collection we will create a square grid and each grid unit will represent the bounds for Foursquare search. 
Each venue (hotel, bus station, food) has a location and category.

Next section will explore the collected data using visualization aids.

In the final section we will use k-means clustering approach to find blobs of location that meet the predefined requirements:
* no other acommodation object in vicinity
* at least one bus station within the walking distance
* at least 5 food places wihthin the walking distance

## Preparation

In [2]:
# imports

# uncomment to install with conda or pip, below is just an example
# !pip install geopy
# !conda install -c conda-forge geopy --yes 

from geopy.geocoders import Nominatim 
from geopy.extra.rate_limiter import RateLimiter

import folium

import requests 
import pandas as pd
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

import numpy as np
import math
import time

import pyproj

In [3]:
# foursquare credentials
foursquare_client_id = 'EXENEWRKIY0ZQPXVOZCD3RWEXTNNQC1113GGBXYXJVHERV0J' # your Foursquare ID
foursquare_client_secret = 'YHAQNOL1EMK0BHONYERC4J52J4ECBPAIQOKWSULTVWZZUQKM' # your Foursquare Secret
foursquare_version = '20180605' # Foursquare API version
foursquare_limit = 100

In [94]:
# Get the Ljubljana tourist center point.

lj_center_address = 'Prešernov trg, Ljubljana, Slovenia'

geolocator = Nominatim(user_agent="LJ_explorer")
geocode = RateLimiter(geolocator.geocode, min_delay_seconds=5)

location = geocode(lj_center_address)
lj_center_lat = location.latitude
lj_center_lon = location.longitude
lj_center_coordinates=(lj_center_lat,lj_center_lon)

print('Coordinates of {} = {}'.format(lj_center_address, lj_center_coordinates))

Coordinates of Prešernov trg, Ljubljana, Slovenia = (46.05140755, 14.506095911950972)


In [95]:
# Helper methods for calculating various geo coordinates in WGS84 space 

geod = pyproj.Geod(ellps='WGS84')

# calculate the geo point which is distance away in the direction of the fwd_azimuth
def calculate_geo_point(start_lat, start_lon, fwd_azimuth, distance):
    end_lon, end_lat, back_azimuth = geod.fwd(start_lon,start_lat,fwd_azimuth,distance)
    return (end_lat, end_lon)

# calculates geo coordinates of corners of a square. Start_lat and start_lon assumes SW corner.
def calculate_square_corners(start_lat, start_lon, square_size):
    corners = [] # clock wise - sw, nw, ne, sw
    corners.append((start_lat, start_lon))
    corners.append(calculate_geo_point(start_lat, start_lon, 0, square_size))
    corners.append(calculate_geo_point(start_lat, start_lon, 45, math.sqrt(2)*square_size))
    corners.append(calculate_geo_point(start_lat, start_lon, 90, square_size))
    return corners

# calculate center coordinates of a square, assumes SW corner
def calculate_square_center(start_lat, start_lon, square_size):
    return calculate_geo_point(start_lat, start_lon, 45, math.sqrt(2)*square_size/2 )

# calculate distance between 2 geo coordinates 
def calculate_distance(start_lat, start_lon, end_lat, end_lon):
    a1, a1, distance = geod.inv(start_lon, start_lat, end_lon, end_lat)
    return distance

In [104]:
# Define the grid = a square defined by coordinates of the sw corner
# Each grid unit/area is a square defined by coordinates of it's corners (sw, nw, ne, se).


def generate_grid(start_lat, start_lon, grid_size, grid_unit):
   
    
    # Calculate the rest of the grid's corners
    #grid_corners = calculate_square_corners(start_lat, start_lon, grid_size)

    # generate the grid
    # start at the SW corner of the grid and make your way up 
    grid=[]
    
    row_origin = (start_lat, start_lon) 
    for i in range(0, math.floor(grid_size/grid_unit)):
        col_origin = row_origin # remember the origin
        for j in range(0, math.floor(grid_size/grid_unit)):
            square = calculate_square_corners(col_origin[0], col_origin[1], grid_unit)
            grid.append(square)
            col_origin = square[3] # new SW corner is the SE corner of the previous square
    
        row_origin = calculate_geo_point(row_origin[0], row_origin[1], 0, grid_unit)
    
    return grid



In [105]:
grid_size = 5000 # meters
grid_unit = 1000 # meters 

# Get the grid's SW corner starting from the Ljubljana center defined by start_lat, start_lon
grid_sw_corner = calculate_geo_point(lj_center_lat, lj_center_lon, 225, grid_size*math.sqrt(2)/2)

grid = generate_grid(grid_sw_corner[0], grid_sw_corner[1], grid_size, grid_unit)
#grid = generate_grid(lj_center_lat, lj_center_lon, grid_size, grid_unit)
print("Number of grid units (areas):", len(grid))

Number of grid units (areas): 25


In [106]:
# Display the grid on the map of Ljubljana together with the center point
map_lj= folium.Map(location=lj_center_coordinates, zoom_start=13)
folium.Marker(lj_center_coordinates, popup=lj_center_address).add_to(map_lj)

for square in grid:
    folium.Rectangle(square, weight=1).add_to(map_lj)

map_lj

## Data collection

In [10]:
# helper functions for pulling the data from Foursquare

radius = 250 # meters

# requires category_id, coordinates of sw corner, coordinates of ne corner
def search_venues_in_area(sw_lat, sw_lon, ne_lat, ne_lon, category_id, client_id, client_secret, version, limit):
    
    url = 'https://api.foursquare.com/v2/venues/search?client_id={}&client_secret={}&v={}&intent=browse&sw={},{}&ne={},{}&categoryId={}&limit={}'.format(
        client_id, 
        client_secret, 
        version,
        sw_lat,
        sw_lon,
        ne_lat,
        ne_lon,
        category_id,    
        limit)
    
    try:
        r = requests.get(url)
        r.raise_for_status()
        return r.json()['response']['venues']
        
    except requests.exceptions.HTTPError as err:
        print(err.response.text)
        return None
    
    
def search_venues_in_radius(lat, lon, category_id, radius, client_id, client_secret, version, limit):
    
    # create the API request URL
    url = 'https://api.foursquare.com/v2/venues/search?client_id={}&client_secret={}&v={}&intent=browse&ll={},{}&categoryId={}&radius={}&limit={}'.format(
        client_id, 
        client_secret, 
        version,
        lat,
        lon,
        category_id,
        radius,
        limit)
    
    try:
        r = requests.get(url)
        r.raise_for_status()
        return r.json()['response']['groups'][0]['items'] 
        
    except requests.exceptions.HTTPError as err:
        print(err.response.text)
        return None

def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']



In [11]:
# Get venues based from grid, limit by category ids

def get_venues_in_grid(categories, grid):
    frames = []
    print("Processing areas: ",end="")
    counter = 0
    for area in grid:
        
        for category_id in categories:
            time.sleep(1)
            
            venues_in_area = search_venues_in_area(area[0][0],
                                                   area[0][1],
                                                   area[2][0],
                                                   area[2][1],
                                                   category_id,
                                                   foursquare_client_id,
                                                   foursquare_client_secret,
                                                   foursquare_version,
                                                   foursquare_limit)
    
            # Check if something went wrong
            if venues_in_area == None:
                print("Search not working...")
                break
        
            # if nothing is found skip to the next area
            if len(venues_in_area) == 0:
                print(".",end="")
                continue
    
            # clean up and generate a panda frame, add to the frames list
            print("+",end="")
            venues_in_area = json_normalize(venues_in_area)
            filtered_columns = ['name', 'categories', 'location.lat', 'location.lng']
            venues_in_area = venues_in_area.loc[:, filtered_columns]
            venues_in_area['categories'] = venues_in_area.apply(get_category_type, axis=1)
            venues_in_area.columns = ['Name', 'Category', 'Latitude', 'Longitude']
            venues_in_area['Area'] = counter
            
            frames.append(venues_in_area)
            ###
        
        counter = counter + 1
        
    # combine results into one frame
    venues = pd.concat(frames, ignore_index=True)

    # remove duplicates (based on venue name) and reset the index
    venues.drop_duplicates(subset ="Name", keep = False, inplace = True)
    venues.reset_index(drop=True, inplace=True)
    
    # add a column for Distance to city center
    venues['Distance to center'] = venues.apply(lambda x: 
                                        calculate_distance(
                                            lj_center_lat, lj_center_lon, 
                                            x["Latitude"], x["Longitude"]), axis=1)
    
    return venues


### Accommodation locations

In [12]:
# Get all the accomodation locations in the grid
hotel_id = ["4bf58dd8d48988d1fa931735"]

hotels = get_venues_in_grid(hotel_id, grid)

Processing areas: ++...+++..+++++.+++++++..

In [16]:
print("The number of acoomodation locations in the grid is:", len(hotels))
hotels.tail(50)

The number of acoomodation locations in the grid is: 144


Name         Category   Latitude  Longitude  \
94   Bed And Breakfast Sincere 1830  Bed & Breakfast  46.052081  14.511585   
95      Holiday Inn Zuerich Airport            Hotel  46.052318  14.503220   
96                Djokic Apartments            Hotel  46.050192  14.505089   
97                 Alibi M14 Hostel           Hostel  46.053416  14.506713   
98                      Munda House           Hostel  46.047733  14.500629   
99       Apartment Under The Castle           Hostel  46.049010  14.511397   
100                      Hotel Park            Hotel  46.053503  14.514570   
101                      DIC hostel           Hostel  46.048119  14.518575   
102                   Hotel Meksiko            Hotel  46.054270  14.519736   
103             Apartment Center 25  Vacation Rental  46.055340  14.514252   
104                 Hostel & SPA 24           Hostel  46.050372  14.513496   
105                      hotel lord       Hotel Pool  46.050529  14.517335   
106                   Meksiko Hotel            Hotel  46.054679  14.519861   
107          Studio Apartment UpArt  Vacation Rental  46.048640  14.513753   
108          Hostel Sleeping Beauty           Hostel  46.052191  14.512714   
109                    Hostel Tabor           Hostel  46.053773  14.513070   
110               Dijaski Dom Tabor           Hostel  46.053684  14.512911   
111                  riverside vila           Resort  46.055055  14.532298   
112                   hostel tivoli           Hostel  46.062221  14.495987   
113             Apartment Pod velbi  Vacation Rental  46.064323  14.494687   
114                     Janez Rooms            Hotel  46.062332  14.495679   
115        Hotel Bellevue Ljubljana            Hotel  46.061692  14.494931   
116              Apartment Colorama  Vacation Rental  46.063739  14.493721   
117      InterContinental Ljubljana            Hotel  46.057172  14.505586   
118             Hotel Lev Ljubljana            Hotel  46.056121  14.502359   
119                   Atelier Hotel            Hotel  46.056978  14.502828   
120                     Airbnb Nina  Bed & Breakfast  46.057210  14.503979   
121            City VIews Apartment  Bed & Breakfast  46.056915  14.504372   
122          Hostel Metro Ljubljana           Hostel  46.055929  14.504070   
123             City View Apartment  Bed & Breakfast  46.056816  14.504374   
124                   Hostel Celica           Hostel  46.056443  14.516538   
125                       TiliaStay            Hotel  46.059692  14.519351   
126                   Hotel Asteria            Hotel  46.056430  14.519132   
127             Budget rooms Vodmat           Hostel  46.057501  14.525004   
128             downtown apartments            Hotel  46.057331  14.517798   
129             Guesthouse Koprivec           Hostel  46.057995  14.527378   
130       Soul Ljubljana Guesthouse  Bed & Breakfast  46.056080  14.537657   
131                Sweet Apartments            Hotel  46.070854  14.485027   
132                   Hostel Aladin           Hostel  46.070946  14.484676   
133          Guesthouse Pri Vodniku            Hotel  46.067350  14.486170   
134                Prenočišča Šiška           Hostel  46.073710  14.485397   
135      Family Guesthouse Greenslo           Hostel  46.070827  14.494866   
136               M Hotel Ljubljana            Hotel  46.068458  14.488867   
137                Pansion Valentin            Hotel  46.065712  14.490685   
138                     hostel 24/1           Hostel  46.072005  14.487920   
139                 The Open Hostel           Hostel  46.072808  14.503842   
140      Julija Bed&Breakfast Hotel            Hotel  46.072186  14.502548   
141                        Vila Ema  Bed & Breakfast  46.068375  14.507369   
142   Holiday Inn Express Ljubljana            Hotel  46.072634  14.503692   
143               MS Sparrow Hostel           Hostel  46.073688  14.510939   

     Area  Distance to center  
94     12          4

### Bus stations/stops locations

In [17]:
# Get all the bus stations in the grid
bus_station_id = ["4bf58dd8d48988d1fe931735", "52f2ab2ebcbc57f1066b8b4f"]

bus_stations = get_venues_in_grid(bus_station_id, grid)


Processing areas: +.+.+.+++.+.+.+.+...+.+.+++.+...+.+++++.+.+.+++.+.

In [18]:
bus_stations.head()

Name     Category   Latitude  Longitude  Area  \
0                   Lpp Mestni Log  Bus Station  46.037838  14.482951     0   
1         LPP postajališče Koprska  Bus Station  46.035342  14.484233     0   
2         LPP postajališče Krimska  Bus Station  46.036959  14.490026     1   
3  Lpp postajalisce Veliki Stradon  Bus Station  46.037190  14.504613     2   
4          LPP postajališče Livada  Bus Station  46.034501  14.517548     3   

   Distance to center  
0         2341.832498  
1         2460.142953  
2         2031.352294  
3         1584.426856  
4         2077.807156

### Food venues locations

In [19]:
# Get all the food venues in the grid

food_id = ["4d4b7105d754a06374d81259"]

food_venues = get_venues_in_grid(food_id, grid)



Processing areas: +++++++++.+++++++++++++++

In [20]:

food_venues.head()
food_venues.describe()

Latitude   Longitude        Area  Distance to center
count  525.000000  525.000000  525.000000          525.000000
mean    46.054238   14.506160   13.560000         1422.102558
std      0.010215    0.014248    6.205969          760.275537
min     46.029031   14.474364    0.000000           78.579787
25%     46.046920   14.496592   10.000000          741.165675
50%     46.053615   14.507094   13.000000         1357.685572
75%     46.062863   14.515011   18.000000         2045.486541
max     46.073877   14.538230   24.000000         3500.635938

In [21]:

# Check numbers of each area
hotels_by_area = hotels.groupby('Area').count()
hotels_by_area

Name  Category  Latitude  Longitude  Distance to center
Area                                                         
0        1         1         1          1                   1
1        2         2         2          2                   2
5        2         2         2          2                   2
6        7         7         7          7                   7
7       23        23        23         23                  23
10       6         6         6          6                   6
11       9         9         9          9                   9
12      50        50        50         50                  50
13      11        11        11         11                  11
14       1         1         1          1                   1
16       5         5         5          5                   5
17       7         7         7          7                   7
18       5         5         5          5                   5
19       2         2         2          2                   2
20       4         4         4          4                   4
21       4         4         4          4                   4
22       5         5         5          5                   5

In [20]:
bus_stations_by_area = bus_stations.groupby('Area').count()
bus_stations_by_area

Name  Category  Latitude  Longitude  Distance to center
Area                                                         
0        2         2         2          2                   2
1        1         1         1          1                   1
2        1         1         1          1                   1
3        3         3         3          3                   3
4        2         2         2          2                   2
5        6         6         6          6                   6
6        6         6         6          6                   6
7        6         6         6          6                   6
8        2         2         2          2                   2
10       1         1         1          1                   1
11       5         5         5          5                   5
12      18        18        18         18                  18
13       8         8         8          8                   8
14       3         3         3          3                   3
16       5         5         5          5                   5
17      13        13        13         13                  13
18       9         9         9          9                   9
19       4         4         4          4                   4
20       6         6         6          6                   6
21       4         4         4          4                   4
22       5         5         5          5                   5
23       6         6         6          6                   6
24       3         3         3          3                   3

In [21]:
food_venues_by_area = food_venues.groupby('Area').count()
food_venues_by_area

Name  Category  Latitude  Longitude  Distance to center
Area                                                         
0       14        14        14         14                  14
1       11        11        11         11                  11
2        3         3         3          3                   3
3        5         5         5          5                   5
4        3         3         3          3                   3
5       22        22        22         22                  22
6       17        17        17         17                  17
7       49        49        49         49                  49
8        7         7         7          7                   7
10      11        11        11         11                  11
11      39        39        39         39                  39
12      46        46        46         46                  46
13      50        50        50         50                  50
14      14        14        14         14                  14
15       1         1         1          1                   1
16      20        20        20         20                  20
17      46        46        46         46                  46
18      48        48        48         48                  48
19      23        23        23         23                  23
20      11        11        11         11                  11
21      25        25        25         25                  25
22      22        22        22         22                  22
23      32        32        32         32                  32
24       6         6         6          6                   6

In [22]:
print("Total number of hotels:", len(hotels))
print("Total number of bus stations:", len(bus_stations))
print("Total number of food venues:", len(food_venues))

Total number of hotels: 144
Total number of bus stations: 119
Total number of food venues: 525


## Analysis

In [152]:
# Visualize accomodation venues, bus stations and food venues on a map

map_lj= folium.Map(location=lj_center_coordinates, zoom_start=13)
folium.Marker(lj_center_coordinates, popup=lj_center_address).add_to(map_lj)
folium.TileLayer('cartodbpositron').add_to(map_lj)
for square in grid:
    folium.Rectangle(square, weight=0.5).add_to(map_lj)
    
for lat, lon, name in zip(hotels['Latitude'], hotels['Longitude'], hotels['Name']):
   
    folium.CircleMarker(
        [lat, lon],
        radius=3,
        color='red',
        fill=True,
        fill_color='red',
        fill_opacity=0.3,
        parse_html=False).add_to(map_lj)  

    
for lat, lon, name in zip(food_venues['Latitude'], food_venues['Longitude'], food_venues['Name']):
   
    folium.CircleMarker(
        [lat, lon],
        radius=3,
        color='blue',
        fill=True,
        fill_color='blue',
        fill_opacity=0.3,
        parse_html=False).add_to(map_lj)  
    
for lat, lon, name in zip(bus_stations['Latitude'], bus_stations['Longitude'], bus_stations['Name']):
   
    folium.CircleMarker(
        [lat, lon],
        radius=3,
        color='yellow',
        fill=True,
        fill_color='yellow',
        fill_opacity=0.3,
        parse_html=False).add_to(map_lj)  
map_lj

In [51]:
# Heatmap of hotel density 

from folium import plugins
from folium.plugins import HeatMap

hotels_coordinates = zip(hotels['Latitude'], hotels['Longitude'])

map_lj= folium.Map(location=lj_center_coordinates, zoom_start=13.6)
#folium.TileLayer('cartodbpositron').add_to(map_lj)
folium.Marker(lj_center_coordinates, popup=lj_center_address).add_to(map_lj)
HeatMap(hotels_coordinates, radius = 20).add_to(map_lj)

for lat, lon in zip(hotels['Latitude'], hotels['Longitude']):
    folium.CircleMarker(
        [lat, lon],
        radius=3,
        color='red',
        fill=True,
        fill_color='red',
        fill_opacity=0.3,
        parse_html=False).add_to(map_lj)  
    
for square in grid:
    folium.Rectangle(square, weight=0.7).add_to(map_lj)
    


map_lj

Here we see that the existing accommodation venues group mostly around the city center in the 500m radius. This is expected. There are also clusters of stretched along the main roads around and inbound to Ljubljana.

There are many pockets of possible locations. 

Now we have all the acommodation venues, bus stations and food venues in scope of few kilometers from hte tourist center of Ljubljana. 

## Analysis

In [ ]:
walking_distance = 250 # meters

In [35]:
# For each hotel get the number of other hotels in walking distance and nearest hotel

def calculate_number_of_nearby_hotels(row):
    count = 0
    for lat, lon in zip(hotels['Latitude'], hotels['Longitude']):
        dist = calculate_distance(row["Latitude"], row["Longitude"], lat, lon)
        if dist > 0 and dist < walking_distance:
            count = count + 1
    return count
###          
    
def calculate_distance_to_nearest_hotel(row):
    min_distance = 100000
    for lat, lon in zip(hotels['Latitude'], hotels['Longitude']):
        dist = calculate_distance(row["Latitude"], row["Longitude"], lat, lon)
        
        if dist > 0 and dist < min_distance:
            min_distance = dist
    return min_distance
###

hotels['Number of nearby hotels'] = hotels.apply(lambda x: calculate_number_of_nearby_hotels(x), axis=1)
hotels['Distance to nearest hotel'] = hotels.apply(lambda x: calculate_distance_to_nearest_hotel(x), axis=1) 




In [112]:
hotels.head()

Name Category   Latitude  Longitude  Area  \
0        Swiss Diamond Hotel    Hotel  46.036586  14.475757     0   
1     AHOTEL Hotel Ljubljana    Hotel  46.030272  14.491635     1   
2   Murgle Luxury Apartments    Hotel  46.035213  14.489431     1   
3  The Vault Hotel Ljubljana    Hotel  46.044758  14.476707     5   
4                Sport Hotel    Hotel  46.038589  14.484196     5   

   Distance to center  Number of nearby hotels  Nearest hotel  \
0         2868.589601                        0     690.208078   
1         2602.324674                        0     575.107919   
2         2214.516878                        0     530.529562   
3         2391.698089                        0     502.633889   
4         2214.345151                        0     429.568496   

   Distance to nearest hotel  Number of nearby bus stations  \
0                 690.208078                              2   
1                 575.107919                              4   
2                 530.529562                              2   
3                 502.633889                              0   
4                 429.568496                              0   

   Distance to nearest bus station  Number of nearby food venues  \
0                        99.628467                             2   
1                       175.917993                             1   
2                       171.214743                             5   
3                       772.679326                             0   
4                       443.903589                             2   

   Distance to nearest food venue  
0                      109.168645  
1                      220.112977  
2                      111.930061  
3                      453.010161  
4                       91.053225

In [37]:
# For each hotel get the number of bus stations in walking distance and nearest bus station

def calculate_number_of_nearby_bus_stations(row):
    count = 0
    for lat, lon in zip(bus_stations['Latitude'], hotels['Longitude']):
        dist = calculate_distance(row["Latitude"], row["Longitude"], lat, lon)
        if dist > 0 and dist < walking_distance:
            count = count + 1
    return count
###

def calculate_distance_to_nearest_bus_station(row):
    min_distance = 100000
    for lat, lon in zip(bus_stations['Latitude'], hotels['Longitude']):
        dist = calculate_distance(row["Latitude"], row["Longitude"], lat, lon)
        
        if dist > 0 and dist < min_distance:
            min_distance = dist
    return min_distance
###

hotels['Number of nearby bus stations'] = hotels.apply(lambda x: calculate_number_of_nearby_bus_stations(x), axis=1)          
hotels['Distance to nearest bus station'] = hotels.apply(lambda x: calculate_distance_to_nearest_bus_station(x), axis=1) 




In [38]:
hotels.head()

Name         Category   Latitude  \
0                          Swiss Diamond Hotel            Hotel  46.036586   
1                       AHOTEL Hotel Ljubljana            Hotel  46.030272   
2                     Murgle Luxury Apartments            Hotel  46.035213   
3                    The Vault Hotel Ljubljana            Hotel  46.044758   
4                                  Sport Hotel            Hotel  46.038589   
5                                 Vila Teslova           Hostel  46.043716   
6                                 Student 2011           Hostel  46.039986   
7                                 SimbolHostel           Hostel  46.039995   
8                               Isabella rooms            Hotel  46.043865   
9          Hotel Sest Pik - Hostel At Six Dots           Hostel  46.044760   
10                                 MartaStudio  Bed & Breakfast  46.045287   
11                                Hotel Sabina            Hotel  46.045899   
12                        Sax Pub & Sax Hostel              Pub  46.043269   
13                                        Trta      Pizza Place  46.043678   
14                                 Adora Hotel   Boarding House  46.045895   
15                              Hotel Galleria            Hotel  46.046755   
16                                Alegro Hotel            Hotel  46.046624   
17                                  Kva Hostel           Hostel  46.046722   
18                 Ljubljana Gardens Apartment            Hotel  46.043461   
19                               Galeria River            Hotel  46.046314   
20                         Slovenika Apartment  Vacation Rental  46.042116   
21                       Antiq Hotel Ljubljana  Bed & Breakfast  46.046767   
22                             Tita Apartments            Hotel  46.038500   
23                             Zois Apartments  Vacation Rental  46.046180   
24                    Cube Central Rooms For 2            Hotel  46.044514   
25                                 Hostel Vrba           Hostel  46.043541   
26                           Hostel Pri Janezu           Hostel  46.044149   
27  Sunflower Accommodation (Pension Soncnica)  Bed & Breakfast  46.040128   
28                          Apartments Hrenova            Hotel  46.044613   
29                       Rocky Mountains House   Boarding House  46.038400   
30                            Aura Guest House  Bed & Breakfast  46.038260   
31                           Viksi Nomad Rooms            Hotel  46.046670   
32                           Confidenti Hostel           Hostel  46.040863   
33                           Nana's Apartments            Hotel  46.045290   
34                       city center apartment  Vacation Rental  46.044610   
35                                Hostel Print           Hostel  46.047046   
36                        Green Fortuna Suites           Hostel  46.049000   
37                 Kitchen Hotel & Hostel Park           Hostel  46.046910   
38                               nana's suites           Resort  46.047040   
39                                 Katrca 1905            Hotel  46.050432   
40                                Nana's rooms           Hostel  46.046985   
41                       Gami Ex-Factory Lofts            Hotel  46.047535   
42                               Vila Veselova           Hostel  46.051122   
43                                       After  Bed & Breakfast  46.047493   
44                                 Rose Valley            Hotel  46.047394   
45                                 Tobacna Red  Vacation Rental  46.047368   
46                Vila Veselova Turquoise Room           Hostel  46.050110   
47                        Sweet Home Ljubljana  Bed & Breakfast  46.047550   
48                            AirBnB By Veljko   Boarding House  46.047553   
49                             Party Pink Room           Hostel  46.051109   

    Longitude  Area  Distance to center  Number of nearby hotel

In [39]:
# For each hotel get the number of food venues in walking distance and nearest food venue
def calculate_number_of_nearby_food_venues(row):
    count = 0
    for lat, lon in zip(food_venues['Latitude'], hotels['Longitude']):
        dist = calculate_distance(row["Latitude"], row["Longitude"], lat, lon)
        if dist > 0 and dist < walking_distance:
            count = count + 1
    return count
###         
    
def calculate_distance_to_nearest_food_venue(row):
    min_distance = 100000
    for lat, lon in zip(food_venues['Latitude'], hotels['Longitude']):
        dist = calculate_distance(row["Latitude"], row["Longitude"], lat, lon)
        
        if dist > 0 and dist < min_distance:
            min_distance = dist
    return min_distance
###

hotels['Number of nearby food venues'] = hotels.apply(lambda x: calculate_number_of_nearby_food_venues(x), axis=1) 
hotels['Distance to nearest food venue'] = hotels.apply(lambda x: calculate_distance_to_nearest_food_venue(x), axis=1) 



In [57]:
hotels.describe()

Latitude   Longitude        Area  Distance to center  \
count  144.000000  144.000000  144.000000          144.000000   
mean    46.051843   14.503039   12.027778         1001.483114   
std      0.008258    0.010507    4.558776          743.197345   
min     46.030272   14.475757    0.000000          104.745374   
25%     46.046927   14.495910   10.000000          464.668797   
50%     46.051115   14.504764   12.000000          718.695752   
75%     46.055376   14.507487   13.000000         1458.612621   
max     46.073710   14.537657   22.000000         2951.348727   

       Number of nearby hotels  Nearest hotel  Distance to nearest hotel  \
count               144.000000     144.000000                 144.000000   
mean                  6.847222     118.425380                 118.425380   
std                   5.966479     137.868372                 137.868372   
min                   0.000000       1.542595                   1.542595   
25%                   2.000000      21.229032                  21.229032   
50%                   5.000000      85.180597                  85.180597   
75%                  11.000000     172.729308                 172.729308   
max                  21.000000     690.208078                 690.208078   

       Number of nearby bus stations  Distance to nearest bus station  \
count                     144.000000                       144.000000   
mean                        3.229167                       223.085426   
std                         2.871901                       259.182267   
min                         0.000000                         0.889218   
25%                         0.000000                        67.947508   
50%                         2.000000                       120.605529   
75%                         6.000000                       267.314624   
max                         9.000000                      1377.895330   

       Number of nearby food venues  Distance to nearest food venue  
count                    144.000000                      144.000000  
mean                       3.944444                      402.201448  
std                        6.940081                      528.649260  
min                        0.000000                        6.705460  
25%                        0.000000                      104.054054  
50%                        1.000000                      208.215993  
75%                        3.250000                      394.463049  
max                       28.000000                     2343.812025

In [58]:
print('Number of hotels without a bus station within the walkign distance', len(hotels[hotels['Number of nearby bus stations'] == 0]))
print('Number of hotels without a food venue within the walking distance', len(hotels[hotels['Number of nearby food venues'] == 0]))

Number of hotels without a bus station within the walkign distance 37
Number of hotels without a food venue within the walkign distance 56


We see that most hotels group together, average distance to nearest hotel is 118 meters. This makes sense especially in the city center. 
Good news is that most hotels have a bus station within the walking distance. However there are 37 hotels that do not have a bus station within the walking distance.

A hotel has on average around 4 food venues within a walking distance. More suprising is that 57 hotels do not have a food venue that is within the walking distance, and that the average distance to a food venue is about 400 meters. Max distance is about 2.3 km. 

In [110]:

#square_center = calculate_square_center(grid[0][0][0], grid[0][0][1], 50)
#subgrid = generate_grid(grid[0][0][0], grid[0][0][1], grid_unit, 50)
#subgrid = generate_grid(lj_center_lat, lj_center_lon, grid_size, grid_unit)
#print("Number of grid units (areas):", len(subgrid))

subgrid_unit = 100 # meters

# for each area create a fine grid that and calculate square centers. This will represent the potential hotel locations.
location_candidates = []
for area in grid:
    subgrid = generate_grid(area[0][0], area[0][1], grid_unit, subgrid_unit)
    for subarea in subgrid:
        subarea_center = calculate_square_center(subarea[0][0], subarea[0][1], subgrid_unit)
        location_candidates.append(subarea_center)


In [123]:
# create a Dataframe with location

#df_locations = pd.DataFrame({'Latitude':[],'Longitude':[]})
df_locations = pd.DataFrame(location_candidates, columns = ['Latitude', 'Longitude'])
#print(len(df_locations))
df_locations.head()

Latitude  Longitude
0  46.029361  14.474452
1  46.029361  14.475743
2  46.029361  14.477035
3  46.029361  14.478326
4  46.029361  14.479618

In [124]:
df_locations['Number of nearby hotels'] = df_locations.apply(lambda x: calculate_number_of_nearby_hotels(x), axis=1)
df_locations['Distance to nearest hotel'] = df_locations.apply(lambda x: calculate_distance_to_nearest_hotel(x), axis=1) 



In [125]:
df_locations['Number of nearby bus stations'] = df_locations.apply(lambda x: calculate_number_of_nearby_bus_stations(x), axis=1)          
df_locations['Distance to nearest bus station'] = df_locations.apply(lambda x: calculate_distance_to_nearest_bus_station(x), axis=1) 



In [126]:

df_locations['Number of nearby food venues'] = df_locations.apply(lambda x: calculate_number_of_nearby_food_venues(x), axis=1) 
df_locations['Distance to nearest food venue'] = df_locations.apply(lambda x: calculate_distance_to_nearest_food_venue(x), axis=1) 




In [142]:
print(len(df_locations))
df_locations.head()

2500


Latitude  Longitude  Number of nearby hotels  Distance to nearest hotel  \
0  46.029361  14.474452                        0                 809.358304   
1  46.029361  14.475743                        0                 803.029279   
2  46.029361  14.477035                        0                 809.103362   
3  46.029361  14.478326                        0                 827.307409   
4  46.029361  14.479618                        0                 856.868669   

   Number of nearby bus stations  Distance to nearest bus station  \
0                              0                       887.582742   
1                              0                       868.684840   
2                              0                       796.077041   
3                              0                       729.973822   
4                              0                       672.296608   

   Number of nearby food venues  Distance to nearest food venue  
0                             0                      706.035100  
1                             0                      615.782972  
2                             0                      529.046813  
3                             0                      447.874028  
4                             0                      375.886307

In [149]:
# Filter out those location that do not have a bus station within a walking distance
# Filter out those locations that do not have a food venue within a walking distance
# Filter out those locations that have a hotel within 400m

loc1 = df_locations[df_locations['Distance to nearest bus station'] < 400]
loc2 = loc1[loc1['Distance to nearest food venue'] < 400]
loc3 = loc2[loc2['Distance to nearest hotel'] > 200]
loc3.reset_index(drop=True, inplace=True)

In [150]:

print(len(loc3))
loc3.head()

322


Latitude  Longitude  Number of nearby hotels  Distance to nearest hotel  \
0  46.029361  14.484784                        0                 539.951272   
1  46.029361  14.486076                        0                 442.124251   
2  46.030261  14.484785                        0                 530.358816   
3  46.030261  14.486076                        0                 430.359975   
4  46.031160  14.483493                        0                 638.029445   

   Number of nearby bus stations  Distance to nearest bus station  \
0                              0                       371.021369   
1                              0                       277.005318   
2                              0                       354.728863   
3                              0                       254.775224   
4                              0                       375.253262   

   Number of nearby food venues  Distance to nearest food venue  
0                             0                      322.607415  
1                             0                      261.406755  
2                             1                      239.429660  
3                             0                      292.672916  
4                             1                       99.389757

In [151]:
# Display the grid on the map of Ljubljana together with the center point
map_lj= folium.Map(location=lj_center_coordinates, zoom_start=13)
folium.Marker(lj_center_coordinates, popup=lj_center_address).add_to(map_lj)

for square in grid:
    folium.Rectangle(square, weight=1).add_to(map_lj)

for lat, lon in zip(loc3['Latitude'], loc3['Longitude']):
    folium.CircleMarker(
        [lat, lon],
        radius=1,
        color='red',
        fill=True,
        fill_color='red',
        fill_opacity=0.3,
        parse_html=False).add_to(map_lj)  
    
map_lj

In [157]:
from sklearn.cluster import KMeans

number_of_clusters = 10

good_coordinates = loc3[['Latitude', 'Longitude']].values
kmeans = KMeans(n_clusters=number_of_clusters, random_state=0).fit(good_coordinates)

cluster_centers = kmeans.cluster_centers_
print(cluster_centers)



[[46.04366015 14.49804506]
 [46.05461252 14.4888708 ]
 [46.04204448 14.5131632 ]
 [46.03289958 14.48512938]
 [46.05248926 14.49474798]
 [46.03379443 14.49444134]
 [46.03768157 14.50584004]
 [46.03736585 14.47809514]
 [46.05445181 14.48252806]
 [46.04089975 14.48950411]]


In [160]:
# Display the grid on the map of Ljubljana together with the center point
map_lj= folium.Map(location=lj_center_coordinates, zoom_start=13)
folium.Marker(lj_center_coordinates, popup=lj_center_address).add_to(map_lj)

for square in grid:
    folium.Rectangle(square, weight=1).add_to(map_lj)

for lat, lon in zip(loc3['Latitude'], loc3['Longitude']):
    folium.CircleMarker(
        [lat, lon],
        radius=1,
        color='red',
        fill=True,
        fill_color='red',
        fill_opacity=0.3,
        parse_html=False).add_to(map_lj)  

    
    

for lat, lon in cluster_centers:
    folium.Circle([lat, lon], 
                  radius=400,
                  color='green', 
                  fill=True, 
                  fill_opacity=0.25).add_to(map_lj) 
    
map_lj